pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go

```diff
@@ -345,39 +359,87 @@ func (r *ReconcileCnsFileAccessConfig) Reconcile(request reconcile.Request) (rec
 	return reconcile.Result{}, nil
 }
 
-func (r *ReconcileCnsFileAccessConfig) configureFileVolumeACL(ctx context.Context, volumeID string, vm *vmoperatortypes.VirtualMachine, instance *cnsfileaccessconfigv1alpha1.CnsFileAccessConfig) error {
+// configureFileVolumeACLWithDelete helps to remove net permissions for file volumes
+// Net permissions will be removed when the external facing IP is the only entry in
+// the ExternalIPtoClientVms list obtained from cnsfilevolumeclient instance
+// For other cases, the method will remove the vm entry from ExternalIPtoClientVms
+//  Returns error if any operation fails
+func (r *ReconcileCnsFileAccessConfig) configureFileVolumeACLWithDelete(ctx context.Context, volumeID string, vm *vmoperatortypes.VirtualMachine, instance *cnsfileaccessconfigv1alpha1.CnsFileAccessConfig) error {
 	log := logger.GetLogger(ctx)
-	networkProvider, err := cnsoperatorutil.GetNetworkProvider(ctx)
+	tkgVMIP, err := r.getExternalIP(ctx, vm)
 	if err != nil {
-		msg := fmt.Sprintf("Failed to identify the network provider. Error: %+v", err)
+		msg := fmt.Sprintf("Failed to get CNSFileVolumeClient instance. Error: %+v", err)
 		log.Error(msg)
 		return errors.New(msg)
 	}
-	var nsxConfiguration bool
-	if networkProvider == "" {
-		return errors.New("Unable to find network provider information")
+	cnsFileVolumeClientInstance, err := cnsfilevolumeclient.GetFileVolumeClientInstance(ctx)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to get CNSFileVolumeClient instance. Error: %+v", err)
+		log.Error(msg)
+		return errors.New(msg)
 	}
-	if networkProvider == cnsoperatorutil.NSXTNetworkProvider {
-		nsxConfiguration = true
-	} else if networkProvider == cnsoperatorutil.VDSNetworkProvider {
-		nsxConfiguration = false
-	} else {
-		msg := fmt.Sprintf("Unknown network provider. Error: %+v", err)
+	clientVms, err := cnsFileVolumeClientInstance.GetClientVMsFromIPList(ctx, instance.Namespace+"/"+instance.Spec.PvcName, tkgVMIP)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to get the list of clients VMs for IP %q. Error: %+v", tkgVMIP, err)
 		log.Error(msg)
 		return errors.New(msg)
 	}
+	if len(clientVms) == 0 {
+		return nil
+	}
+	if len(clientVms) == 1 && clientVms[0] == vm.Name {
+		// Prepare the CnsVolumeACLConfigureSpec
+		cnsVolumeID := cnstypes.CnsVolumeId{
```

Comment:
Can we move this logic of creating spec to a helper method, and use that method while adding or removing permissions by passing a bool? As far as I see, setting `Delete:  true` in `cnsNFSAccessControlSpecList` is only diff while creating the spec for add or delete.

Reply:
+1

Reply:
Done